cpp/src/copying/copy_range.cu

```diff
@@ -0,0 +1,167 @@
+
+namespace {
+
+struct column_range_factory {
+  cudf::column_view input;
+  cudf::size_type offset;
+
+  template <typename T>
+  struct column_range {
+    T const* column_data;
+    cudf::bitmask_type const* mask;
+    cudf::size_type begin;
+
+    __device__
+    T data(cudf::size_type index) { 
+      return column_data[begin + index]; }
+
+    __device__
+    bool valid(cudf::size_type index) {
+      return cudf::util::bit_is_set(mask, begin + index);
+    }
+  };
+
+  template <typename T>
+  column_range<T> make() {
+    return column_range<T>{
+      input.head<T>(),
+      input.null_mask(),
+      input.offset() + offset,
+    };
+  }
+};
+
+}
+
+namespace cudf {
+namespace experimental {
+
+namespace detail {
+
+void copy_range(mutable_column_view& output, column_view const& input,
+                size_type out_begin, size_type out_end, size_type in_begin,
+                cudaStream_t stream) {
+  CUDF_EXPECTS(cudf::is_fixed_width(output.type()) == true,
+               "In-place copy_range does not support variable-sized types.");
+  CUDF_EXPECTS((out_begin >= 0) &&
+               (out_begin <= out_end) &&
+               (out_begin < output.size()) &&
+               (out_end <= output.size()) &&
+               (in_begin >= 0) &&
+               (in_begin < input.size()) &&
+               (in_begin + out_end - out_begin <= input.size()),
+               "Range is out of bounds.");
+  CUDF_EXPECTS(output.type() == input.type(), "Data type mismatch.");
+  CUDF_EXPECTS((output.nullable() == true) || (input.has_nulls() == false),
+               "output should be nullable if input has null values.");
+
+  if (out_end != out_begin) {  // otherwise no-op
+    copy_range(
+      output,
+      column_range_factory{input, in_begin},
+      out_begin, out_end, stream);
+  }
+}
+
+std::unique_ptr<column> copy_range(column_view const& output,
+                                   column_view const& input,
+                                   size_type out_begin, size_type out_end,
+                                   size_type in_begin,
+                                   cudaStream_t stream,
+                                   rmm::mr::device_memory_resource* mr) {
+  CUDF_EXPECTS(cudf::is_fixed_width(output.type()) == true,
+               "Variable-sized types are not supported yet.");
+  CUDF_EXPECTS((out_begin >= 0) &&
+               (out_begin <= out_end) &&
+               (out_begin < output.size()) &&
+               (out_end <= output.size()) &&
+               (in_begin >= 0) &&
+               (in_begin < input.size()) &&
+               (in_begin + out_end - out_begin <= input.size()),
+               "Range is out of bounds.");
+  CUDF_EXPECTS(output.type() == input.type(), "Data type mismatch.");
+
+  auto state = mask_state{UNALLOCATED};
+  if (input.has_nulls() == true) {
+    state = UNINITIALIZED;
+  }
+
+  auto ret = std::unique_ptr<column>{nullptr};
+
+  if (cudf::is_numeric(output.type()) == true) {
```

Comment:
I think you can just use `allocate_like` here.

Reply:
Probably.